paymentsheet/src/main/java/com/stripe/android/customersheet/CustomerSheetViewModel.kt

```diff
@@ -326,12 +327,83 @@ internal class CustomerSheetViewModel @Inject constructor(
-        updateViewState<CustomerSheetViewState.AddPaymentMethod> {
-            it.copy(isProcessing = false)
+        viewModelScope.launch {
+            if (customerAdapter.canCreateSetupIntents) {
+                attachWithSetupIntent(paymentMethod = paymentMethod)
+            } else {
+                attachPaymentMethod(paymentMethod = paymentMethod)
+            }
+        }
+    }
+
+    private suspend fun attachWithSetupIntent(paymentMethod: PaymentMethod) {
+        customerAdapter.setupIntentClientSecretForCustomerAttach()
+            .onSuccess { clientSecret ->
+                try {
+                    val setupIntent = stripeRepository.confirmSetupIntent(
+                        confirmSetupIntentParams = ConfirmSetupIntentParams.create(
+                            paymentMethodId = paymentMethod.id!!,
+                            clientSecret = clientSecret,
+                        ),
+                        options = ApiRequest.Options(
+                            apiKey = paymentConfiguration.publishableKey,
+                            stripeAccount = paymentConfiguration.stripeAccountId,
+                        ),
+                    )
+                    setupIntent?.let {
+                        handlePaymentMethodAttachSuccess(paymentMethod)
+                    }
+                } catch (ex: Exception) {
+                    // TODO (jameswoo) translate error message
+                    updateViewState<CustomerSheetViewState.AddPaymentMethod> {
+                        it.copy(
+                            errorMessage = "Unable to attach this payment method",
+                            isProcessing = false,
+                        )
+                    }
+                }
+            }.onFailure {
+                // TODO (jameswoo) translate error message
+                updateViewState<CustomerSheetViewState.AddPaymentMethod> {
+                    it.copy(
+                        errorMessage = "Unable to attach this payment method",
+                        isProcessing = false,
+                    )
+                }
+            }
+    }
+
+    private suspend fun attachPaymentMethod(paymentMethod: PaymentMethod) {
+        customerAdapter.attachPaymentMethod(paymentMethod.id!!)
+            .onSuccess {
+                handlePaymentMethodAttachSuccess(paymentMethod)
+            }
+            .onFailure {
+                // TODO (jameswoo) translate error message
+                updateViewState<CustomerSheetViewState.AddPaymentMethod> {
+                    it.copy(
+                        errorMessage = "Unable to attach this payment method",
+                        isProcessing = false,
+                    )
+                }
+            }
+    }
+
+    private fun handlePaymentMethodAttachSuccess(paymentMethod: PaymentMethod) {
+        onBackPressed()
+        updateViewState<CustomerSheetViewState.SelectPaymentMethod> {
```

Comment:
Can we just pop the backstack in the `updateViewState` closure?

Also, what do you think about only having the backstack in the view state instead of as a separate class member?

Reply:
`updateViewState` is meant to update the existing view state. I don't know if it makes sense to pop the backstack if we are updating the current screen.

> Also, what do you think about only having the backstack in the view state instead of as a separate class member?
What makes this better?